src/shared/openssl-util.c

```diff
@@ -1125,6 +1125,83 @@ int digest_and_sign(
+
+        int x509_mdnid = 0, x509_pknid = 0;
+        if (X509_get_signature_info(certificate, &x509_mdnid, &x509_pknid, NULL, NULL) == 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to get X509 digest NID: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        const EVP_MD *md = EVP_get_digestbynid(x509_mdnid);
+        if (!md)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to get digest algorithm via digest NID");
+
+        _cleanup_(PKCS7_SIGNER_INFO_freep) PKCS7_SIGNER_INFO *si = PKCS7_SIGNER_INFO_new();
+        if (!si)
+                return log_oom();
+
+        if (private_key) {
+                if (PKCS7_SIGNER_INFO_set(si, certificate, private_key, EVP_get_digestbynid(x509_mdnid)) <= 0)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to configure signer info: %s",
+                                               ERR_error_string(ERR_get_error(), NULL));
+        } else {
+                if (ASN1_INTEGER_set(si->version, 1) == 0)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set signer info version: %s",
+                                               ERR_error_string(ERR_get_error(), NULL));
+
+                if (X509_NAME_set(&si->issuer_and_serial->issuer, X509_get_issuer_name(certificate)) == 0)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set signer info issuer: %s",
+                                               ERR_error_string(ERR_get_error(), NULL));
+
+                ASN1_INTEGER_free(si->issuer_and_serial->serial);
+                si->issuer_and_serial->serial = ASN1_INTEGER_dup(X509_get0_serialNumber(certificate));
+                if (!si->issuer_and_serial->serial)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set signer info serial: %s",
+                                               ERR_error_string(ERR_get_error(), NULL));
+
+                if (X509_ALGOR_set0(si->digest_alg, OBJ_nid2obj(x509_mdnid), V_ASN1_NULL, NULL) == 0)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set signer info digest algorithm: %s",
+                                               ERR_error_string(ERR_get_error(), NULL));
+
+                if (X509_ALGOR_set0(si->digest_enc_alg, OBJ_nid2obj(x509_pknid), V_ASN1_NULL, NULL) == 0)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set signer info signing algorithm: %s",
+                                               ERR_error_string(ERR_get_error(), NULL));
+        }
+
+        if (PKCS7_add_signer(p7, si) == 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set PKCS7 signer info: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        *ret_p7 = TAKE_PTR(p7);
+        if (ret_si)
+                *ret_si = TAKE_PTR(si);
```

Comment:
Not a big matter, but `TAKE_PTR()` here is redundant.

Reply:
yeah, please drop the TAKE_PTR(). it suggests this was about ownership, but really isn't.

maybe also add brief comment here: `/* NB: we do not pass ownership here, 'si' object remains owned by 'p7' object */

Reply:
can we make `ret_si` param const? to indicate that this is foreign owned? (not sure openssl allows that, but if so, would be good)